supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -59,6 +61,119 @@ case object Cancelled extends TesRunStatus {
+       |
+       |# Remove the leading and trailing quotes
+       |BEARER_TOKEN="$${BEARER_TOKEN#\\"}"
+       |BEARER_TOKEN="$${BEARER_TOKEN%\\"}"
+       |
+       |# Use the precomputed endpoint from cromwell + WSM to acquire a sas token
+       |echo Requesting sas token from WSM...
+       |sas_response_json=$$(curl -s \\
+       |                    --retry 3 \\
+       |                    --retry-delay 2 \\
+       |                    -X POST "$getSasWsmEndpoint" \\
+       |                    -H "Content-Type: application/json" \\
+       |                    -H "accept: */*" \\
+       |                    -H "Authorization: Bearer $${BEARER_TOKEN}")
+       |
+       |# Store token as environment variable
+       |export $environmentVariableName=$$(echo "$${sas_response_json}" | jq -r '.token')
+       |
+       |# Echo the first characters for logging/debugging purposes. "null" indicates something went wrong.
+       |echo Saving sas token: $${$environmentVariableName:0:4}**** to environment variable $environmentVariableName...
+       |### END ACQUIRE LOCAL SAS TOKEN ###
+       |""".stripMargin
+  }
+
+  private def maybeConvertToBlob(pathToTest: Try[Path]): Try[BlobPath] = {
+    pathToTest.collect { case blob: BlobPath => blob }
+  }
+
+  /**
+   * Under certain situations (and only on Terra), we want the VM running a TES task to have the ability to acquire a
+   * fresh sas token for itself. In order to be able to do this, we provide it with a precomputed endpoint it can use.
+   * The task VM will use the user assigned managed identity that it is running as in order to authenticate.
+   * We only return a value if at least one of the inputs is a BlobPath and //TODO flag specified in WDL.
+   * @param taskInputs The inputs to this particular TesTask. If any are blob files, the first  will be used to
+   *                   determine the storage container to retrieve the sas token for.
+   * @param pathGetter A function to convert string filepath into a cromwell Path object.
+   * @param blobConverter A function to convert a Path into a Blob path, if possible. Provided for testing purposes.
+   * @return A URL endpoint that, when called with proper authentication, will return a sas token.
+   *         Returns 'None' if one should not be used for this task.
+   */
+  def determineWSMSasEndpointFromInputs(taskInputs: List[Input],
+                                        pathGetter: String => Try[Path],
+                                        logger: JobLogger,
+                                        blobConverter: Try[Path] => Try[BlobPath] = maybeConvertToBlob): Try[String] = {
+    // Collect all of the inputs that are valid blob paths
+    val blobFiles = taskInputs.collect{
+      case Input(_, _, Some(url), _, _, _) => BlobPathBuilder.validateBlobPath(url)
```

Comment:
Can you do the conversion to `BlobPath` (or not) here, so we don't need to deal with the validation pieces directly? Something like
```
val blobFiles = taskInputs.collect {
  case Input(_, _, Some(url), _, _, _) => blobConverter(pathGetter(url))
}.collect {
  case Success(blob) => blob
}
```